### ISSUE TITLE
matrix dimensions transposed

### ISSUE THREAD

**Issue by mtmorgan** [mojaveazure/loomR#1]
It's confusing, but I think matrix data is read / written transposed from native hdf5 format
```
> create(fl <- tempfile(), matrix(1:6, 3))
Class: loom
Filename: /tmp/RtmpuMT4OX/file308568716c66
Access type: H5F_ACC_RDWR
Attributes: version, chunks
Listing:
      name      object.type dataset.dims dataset.type_class
 col_attrs   H5O_TYPE_GROUP         <NA>               <NA>
    layers   H5O_TYPE_GROUP         <NA>               <NA>
    matrix H5O_TYPE_DATASET        3 x 2        H5T_INTEGER
 row_attrs   H5O_TYPE_GROUP         <NA>               <NA>
> system2("h5ls", fl)
col_attrs                Group
layers                   Group
matrix                   Dataset {2/Inf, 3/Inf}
row_attrs                Group
```

Similarly with [test_loom.loom][1]
```
> connect("test_loom.loom")
Class: loom
Filename: /home/mtmorgan/Downloads/test_loom.loom
Access type: H5F_ACC_RDONLY
Attributes: creation_date, chunks
Listing:
      name      object.type dataset.dims dataset.type_class
 col_attrs   H5O_TYPE_GROUP         <NA>               <NA>
    layers   H5O_TYPE_GROUP         <NA>               <NA>
    matrix H5O_TYPE_DATASET    2100 x 21          H5T_FLOAT
 row_attrs   H5O_TYPE_GROUP         <NA>               <NA>
> x = connect("test_loom.loom")
> dim(x$matrix$read())
[1] 2100   21
> system2("h5ls", "test_loom.loom")
col_attrs                Group
layers                   Group
matrix                   Dataset {21, 2100/Inf}
row_attrs                Group
```

I guess this is an issue in hdf5r that needs to be fixed or accomodated in loomR; rhdf5 has the same issue

```
> dim(rhdf5::h5read("test_loom.loom", "matrix"))
[1] 2100   21
> length(rhdf5::h5read("test_loom.loom", "row_attrs")$Gene)
[1] 21
```

[1]: https://github.com/linnarsson-lab/loom-viewer/blob/master/test_loom.loom

**Comment by mojaveazure** [comment]
The transpose issue is an R issue, not an rdhf5 or hdf5r issue. I'll admit, I don't fully understand the issue, but it's something to do with column-major vs. row-major array access. For efficiency of access, both of these HDF5 libraries use R's (Fortran's) column-major access compared to Python's (C's) row-major. This is why Python HDF5 libraries view matrices in a "normal" orientation while loomR views them as "transposed"; despite this, they are, on disk, unchanged. Using the test loom file you linked to, we can view the shape in both loompy and loomR.

```python
>>> import loompy
>>> lfile = loompy.connect('test_loom.loom', 'r')
>>> lfile.shape
(21, 2100)
```

```R
> lfile <- loomR::connect('test_loom.loom')
> lfile$shape
[1] 2100   21
```

More information about this row-major vs column-major stuff can be found [here](https://github.com/hhoeflin/hdf5r/issues/81#issuecomment-340180514) and [here](https://cran.r-project.org/web/packages/reticulate/vignettes/arrays.html).

I have tried to make this clear in loomR's documentation, and have stated that `loomR::create` requires the data matrix to have cells as rows and cells as columns. I have made this choice to follow the philosophy of making as few assumptions, and to not confuse users when they access data from the matrix and get cells by genes instead of genes by cells. The argument could be made that I could rely on the `gene.attrs` and `cell.attrs` arguments, but that would assume the user provides this data.

I am open to this debate, as to whether `loomR::create` should transpose the data for the user and assume the user knows this vs not transposing the data and laying out in the documentation that `loomR::create` requires cells by genes. Thoughts?

**Comment by mtmorgan** [comment]
I think rows should be rows and columns columns; this is reinforced by row_attrs and col_attrs of the loom format. It's not an R issue; R represents data one way, consistently, and I guess the rhdf5 / hdf5r authors opted (explicitly, in the hdf5r link you provide) for efficiency over portability, which doesn't really make sense when adopting a file format for interoperability reasons (and when the cost of transposition, implemented at the C level, will be small compared to operations that one might do, e.g., at the R level). The user shouldn't have to worry about such complicated issues -- they just want their row x column matrix in R to be conceptually the same row x column matrix their friend sees in python.

As an aside, C doesn't have a native array representation, so it doesn't really make sense to talk about how C represents arrays; R allocates a vector and calculates offsets into it  `(i - 1) + (j -  1) * nrow`. The transpose representation would have been `(i - 1) * ncol + (j - 1)` (unless I got that backward ...), some C code allocates a vector and pointers into it `x[i][j]`, which is the transpose of Forrtan, but again that's an implementation on top of native C.

Of course all just my two cents...

**Comment by ChristophH** [comment]
I agree with mtmorgan. loomR shouldn't make any assumptions on whether rows are genes or cells. but it should make sure that a matrix in a loom file is in the same orientation regardless whether it is opened/created in loomR or loompy. The loom file specification is quite general (there is no mention of cells or genes).
